### ISSUE TITLE
Linkerd controller pods are not coming.

### ISSUE THREAD

**Issue by sathishkumar3009** [linkerd/linkerd2#7938]
### What is the issue?

Deployed linkerD using helm charts.
Repository: "https://helm.linkerd.io/stable"
Version: 2.10.0

Pods of linkerd is not coming up due to the below error. 
[  3115.012070s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN

NAME                                      READY        STATUS       RESTARTS   AGE
linkerd-controller-6568cb69c4-j6j55          0/2             CrashLoopBackOff   17         50m
linkerd-destination-7f4b5bfd-8kztt            0/2             CrashLoopBackOff   17         50m
linkerd-identity-64894647c9-7jkmh           0/2             Init:0/1           0          50m
linkerd-proxy-injector-5d45fb6b6b-zkpsl   0/2            Init:0/1           0          50m
linkerd-sp-validator-7ddb4b68f4-xjztl        0/2            Init:0/1           0          50m

NAME                        TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)    AGE
linkerd-controller-api      ClusterIP   10.0.166.165   <none>        8085/TCP   50m
linkerd-dst                 ClusterIP   10.0.251.17    <none>        8086/TCP   50m
linkerd-dst-headless        ClusterIP   None           <none>        8086/TCP   50m
linkerd-identity            ClusterIP   10.0.145.50    <none>        8080/TCP   50m
linkerd-identity-headless   ClusterIP   None           <none>        8080/TCP   50m
linkerd-proxy-injector      ClusterIP   10.0.75.48     <none>        443/TCP    50m
linkerd-sp-validator        ClusterIP   10.0.101.116   <none>        443/TCP    50m

### How can it be reproduced?

Deploying helm charts from this repo.https://github.com/Azure-Terraform/terraform-helm-linkerd/blob/main/variables.tf

### Logs, error output, etc

[  3115.012070s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN

### output of `linkerd check -o short`

Notworking

### Environment

Kubernetes version 1.22.0

**Comment by mateiidavid** [comment]
Hey @sathishkumar3009 

> [ 3115.012070s] WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN

Is this the only log warning you're seeing? This appears to be recorded around 50 minutes after your pod's proxy was started (i.e `3115.012070s`). In order for the proxy to function correctly, it needs to establish a connection to the identity service. Can you confirm the identity pod for linkerd is up and running? (`kubectl get pods -n linkerd`).

It would be helpful to see if the services have been created: `kubectl get services -n linkerd`.

And finally, can you confirm your k8s cluster has successfully set-up a DNS record for the headless identity service? I find that `dig SRV <host>` works best here, but you could also look-up the hostname using `nslookup`. The Kubernetes official docs have a guide on how to [debug DNS resolution](https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/).

**Comment by sathishkumar3009** [comment]
Hey david, thanks for the response,

Now all the pods are up and running but getting different warnings in the logs of Identity and controller. Is this an issue?

Identity Logs:
 kubectl logs linkerd-identity-74c57b7fdb-pm4zh -n linkerd -c linkerd-proxy
time="2022-02-22T13:47:23Z" level=info msg="running version stable-2.10.1"
[     0.004617s]  INFO ThreadId(01) linkerd2_proxy::rt: Using single-threaded proxy runtime
[     0.005654s]  INFO ThreadId(01) linkerd2_proxy: Admin interface on 0.0.0.0:4191
[     0.005681s]  INFO ThreadId(01) linkerd2_proxy: Inbound interface on 0.0.0.0:4143
[     0.005689s]  INFO ThreadId(01) linkerd2_proxy: Outbound interface on 127.0.0.1:4140
[     0.005696s]  INFO ThreadId(01) linkerd2_proxy: Tap DISABLED
[     0.005702s]  INFO ThreadId(01) linkerd2_proxy: Local identity is linkerd-identity.linkerd.serviceaccount.identity.linkerd.cluster.local
[     0.005719s]  INFO ThreadId(01) linkerd2_proxy: Identity verified via localhost:8080
[     0.005727s]  INFO ThreadId(01) linkerd2_proxy: Destinations resolved via linkerd-dst-headless.linkerd.svc.cluster.local:8086 (linkerd-destination.linkerd.serviceaccount.identity.linkerd.cluster.local)
[     0.029063s]  INFO ThreadId(02) daemon:identity: linkerd_app: Certified identity: linkerd-identity.linkerd.serviceaccount.identity.linkerd.cluster.local
[     9.935615s]  WARN ThreadId(01) server{port=8080}: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-dst-headless.linkerd.svc.cluster.local. type: SRV class: IN
[     9.938528s]  WARN ThreadId(01) server{port=8080}: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-dst-headless.linkerd.svc.cluster.local. type: SRV class: IN
[    14.937915s]  WARN ThreadId(01) server{port=8080}: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-dst-headless.linkerd.svc.cluster.local. type: SRV class: IN
[    19.938312s]  WARN ThreadId(01) server{port=8080}: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-dst-headless.linkerd.svc.cluster.local. type: SRV class: IN


Controller Logs: linkerd-dst-headless.linkerd.svc.cluster.local. type: SRV class: IN
[    10.618514s]  WARN ThreadId(01) server{orig_dst=10.0.244.53:8086}: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-dst-headless.linkerd.svc.cluster.local. type: SRV class: IN
[    13.616348s]  WARN ThreadId(01) server{orig_dst=10.0.244.53:8086}: linkerd_timeout::failfast: TCP Server entering failfast after 3s
[    13.616498s]  INFO ThreadId(01) linkerd_app_core::serve: Connection closed error=TCP Server service in fail-fast
[    14.633118s]  INFO ThreadId(01) linkerd_app_core::serve: Connection closed error=TCP Server service in fail-fast
[    15.618967s]  WARN ThreadId(01) server{orig_dst=10.0.244.53:8086}: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-dst-headless.linkerd.svc.cluster.local. type: SRV class: IN
[    16.249476s]  INFO ThreadId(01) linkerd_app_core::serve: Connection closed error=TCP Server service in fail-fast
[    18.944873s]  INFO ThreadId(01) linkerd_app_core::serve: Connection closed error=TCP Server service in fail-fast
[    20.618547s]  WARN ThreadId(01) server{orig_dst=10.0.244.53:8086}: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-dst-headless.linkerd.svc.cluster.local. type: SRV class: IN
[    22.337432s]  INFO ThreadId(01) linkerd_app_core::serve: Connection closed error=TCP Server service in fail-fast

**Comment by mateiidavid** [comment]
If the pods have _all_ successfully come up but you still get warnings, it means your DNS resolution might not work correctly in the cluster. You can follow the steps I posted above to pinpoint the issue:

> It would be helpful to see if the services have been created: kubectl get services -n linkerd.
>
> And finally, can you confirm your k8s cluster has successfully set-up a DNS record for the headless identity service? I find that dig SRV <host> works best here, but you could also look-up the hostname using nslookup. The Kubernetes official docs have a guide on how to [debug DNS resolution](https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/).

**Comment by sathishkumar3009** [comment]
Output of kubectl get svc -n linkerd

NAME                                TYPE        CLUSTER-IP    EXTERNAL-IP   PORT(S)    AGE
linkerd-controller-api      ClusterIP   10.0.13.217   <none>        8085/TCP   23h
linkerd-dst                       ClusterIP   10.0.244.53   <none>        8086/TCP   23h
linkerd-dst-headless        ClusterIP   None          <none>        8086/TCP   23h
linkerd-identity                ClusterIP   10.0.43.176   <none>        8080/TCP   23h
linkerd-identity-headless ClusterIP   None          <none>        8080/TCP   23h
linkerd-proxy-injector      ClusterIP   10.0.180.79   <none>        443/TCP    23h
linkerd-sp-validator        ClusterIP   10.0.107.59   <none>        443/TCP    23h

**Comment by mateiidavid** [comment]
Have you managed to run `dig` or `nslookup` on the headless host (`linkerd-dst-headless.linkerd.svc.cluster.local`)?

**Comment by sathishkumar3009** [comment]
I have followed the above mentioned page and got this as output.
kubectl exec -i -t dnsutils -- nslookup linkerd-dst-headless.linkerd.svc.cluster.local
Server:         10.0.0.10
Address:        10.0.0.10#53

Name:   linkerd-dst-headless.linkerd.svc.cluster.local
Address: 192.168.2.185

**Comment by sathishkumar3009** [comment]
@mateiidavid Please help me
    I have injected linkerd for few of our services like vault and jenkins and there is no issue.

   And I tried injected for Jfrog Artifactory, Pods are not at all coming up for services and also linkerd-proxy containers.

   Below is the describe of pod and linkerd init container is getting failed.
    
   Reason:  iptables v1.8.2 (legacy): can't initialize iptables table `nat': Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded.

  linkerd-init:
    Container ID:  containerd://31192fc213ef1d26e8726fc3facf284d6458b4a19eb826cd115c3d45a40e8358
    Image:         cr.l5d.io/linkerd/proxy-init:v1.3.11
    Image ID:      cr.l5d.io/linkerd/proxy-init@sha256:9a7d17d2bce7274f5abfd611d34e24460a59c55e55ff93440c31ab6b60b542b2
    Port:          <none>
    Host Port:     <none>
    Args:
      --incoming-proxy-port
      4143
      --outgoing-proxy-port
      4140
      --proxy-uid
      2102
      --inbound-ports-to-ignore
      4190,4191
    State:       Waiting
      Reason:    CrashLoopBackOff
    Last State:  Terminated
      Reason:    Error
      Message:   /02/25 08:33:56 Will ignore port [4190 4191] on chain PROXY_INIT_REDIRECT
2022/02/25 08:33:56 Will redirect all INPUT ports to proxy
2022/02/25 08:33:56 Ignoring uid 2102
2022/02/25 08:33:56 Redirecting all OUTPUT to 4140
2022/02/25 08:33:56

2022/02/25 08:33:56 adding rules
------------------------------------------------------------
2022/02/25 08:33:56 :; iptables -t nat -N PROXY_INIT_REDIRECT -m comment --comment proxy-init/redirect-common-chain/1645778036
2022/02/25 08:33:56 iptables v1.8.2 (legacy): can't initialize iptables table `nat': Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded.

2022/02/25 08:33:56 Aborting firewall configuration
Error: exit status 3

**Comment by mateiidavid** [comment]
Hey @sathishkumar3009, can you try to upgrade Linkerd or re-install and set the following flag? `--set "proxyInit.runAsRoot=true"`. It seems that iptables needs to run as root to be able to set things up.

**Comment by sathishkumar3009** [comment]
@mateiidavid 

   Yes we are using AKS cluster and deployed all the services in it using helm charts.

   How do I input below values in the Helm charts? Please suggest. Should I pass this value under securityContext of helm chart? I'm using stable version of linkerd 2.10.1

   --set "proxyInit.runAsRoot=true"

**Comment by sathishkumar3009** [comment]
@mateiidavid 

     I have set that value in helm and it is working now. Thanks. Jfrog application is also working as expected.

    But I could see these logs in linkerd-proxy container. Any solution for this? For other applications logs are clear.

   [     1.050594s]  INFO ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}: linkerd_app_core::serve: Connection closed error=non-HTTP connection from 127.0.0.1:51526
[    31.968431s]  INFO ThreadId(01) inbound:server{port=8082}:rescue{client.addr=10.1.0.67:58104}: linkerd_app_core::errors::respond: Request failed error=error trying to connect: Connection refused (os error 111)
[    36.967696s]  INFO ThreadId(01) inbound:server{port=8082}:rescue{client.addr=10.1.0.67:58156}: linkerd_app_core::errors::respond: Request failed error=error trying to connect: Connection refused (os error 111)

**Comment by sathishkumar3009** [comment]
@mateiidavid  

     Yes we are using AKS managed cluster and linkerd is deployed using helm chart.

**Comment by sathishkumar3009** [comment]
@mateiidavid 

     Somehow I have brought up all my services and all the pods are meshed.

     At last i installed linkerd as well. Able to see the stat for my deployments and no issue in it.

     But when I try to see linkerd viz dashboard &    unable to see the UI using this URL. http://localhost:50750/

      It gives me connection refused error.

**Comment by sathishkumar3009** [comment]
@mateiidavid Similar issue was raised but no solution has been provided . https://github.com/linkerd/linkerd2/issues/6248
